src/utf8.cpp

```diff
@@ -0,0 +1,346 @@
+                const utf8proc_property_t *codepointProperty = utf8proc_get_property(wbBuffer[wbIndex]);
+
+                // If this is a uppercase codepoint, we do a tolower then toupper
+                if (codepointProperty->category == UTF8PROC_CATEGORY_LU) {
+                    wbBuffer[wbIndex] = utf8proc_toupper(utf8proc_tolower(wbBuffer[wbIndex]));
+                }
+
+                // If this is a lowercase codepoint, we do a toupper then tolower
+                else if (codepointProperty->category == UTF8PROC_CATEGORY_LL) {
+                    wbBuffer[wbIndex] = utf8proc_tolower(utf8proc_toupper(wbBuffer[wbIndex]));
+                }
+            }
+        }
+    }
+
+    return decomposedLength;
+}
+
+bool normalize_string(char **_utf8Buffer, uint64_t &bufferLength)
+{
+    const char *utf8Buffer = *_utf8Buffer;
+    int32_t defaultInFlightBuffer[128], *inFlightBuffer = defaultInFlightBuffer;
+    size_t inFlightBufferLength = 128;
+    std::vector<ScratchpadChunck> scratchPad;
+
+    // A tricky part of this conversion is that the output size is totally unknown, but we want to be efficient with our
+    // allocations. We're going to write the glyph we're normalising in a static buffer (if possible) and write the
+    // filtered, normalized results in a bunch of semi-fixed buffer (the scratchpad) Only when the conversion is over
+    // will we allocate the final buffer and copy everything in there.
+    scratchPad.reserve(8);
+    scratchPad.emplace_back(bufferLength > 1024 ? bufferLength : 1024);
+
+    uint32_t codepoint;
+    uint64_t position = 0;
+    while ((codepoint = fetch_next_codepoint(utf8Buffer, position, bufferLength)) != UTF8_EOF) {
+        // Ignore invalid glyphs
+        if (codepoint == UTF8_INVALID) {
+            continue;
+        }
+
+        size_t decomposedLength = normalize_codepoint(codepoint, inFlightBuffer, inFlightBufferLength);
+
+        if (decomposedLength > inFlightBufferLength) {
+            // We have to increase the size of the in-flight buffer :(
+            if (inFlightBuffer != defaultInFlightBuffer) {
+                free(inFlightBuffer);
+            }
+            inFlightBufferLength = decomposedLength + 64;
+            inFlightBuffer = (int32_t *)malloc((size_t)inFlightBufferLength * sizeof(int32_t));
+            if (inFlightBuffer == nullptr) {
+                return false;
+            }
+
+            // Run the decomposition again with the larger buffer.
+            decomposedLength = normalize_codepoint(codepoint, inFlightBuffer, inFlightBufferLength);
+        }
+
+        // Write the codepoints to the scratchpad
+        for (size_t inflightBufferIndex = 0; inflightBufferIndex < decomposedLength; ++inflightBufferIndex) {
+            char utf8Write[4];
+            uint8_t lengthWritten = write_codepoint((uint32_t)inFlightBuffer[inflightBufferIndex], utf8Write, 4);
+
+            if (scratchPad.back().used + lengthWritten >= scratchPad.back().length) {
```

Comment:
Instead of `scratchPad.back().length` this should be a fixed size, as if the input string is small but the results is much larger, it will result in many small allocations.

Reply:
The chunks are a minimum of 1kB long to handle this specific case :)

Reply:
Ah right, that should work fine then!